src/components/Carousel/index.tsx

```diff
@@ -0,0 +1,234 @@
+import { Autoplay, Pagination } from 'swiper/dist/js/swiper.esm';
+
+import { media, SIZES } from '../../BreakPoints';
+import { gray700, white } from '../../Colors';
+import Navigation, { CarouselNavigationPosition } from './Navigation';
+import ViewAllButton from './ViewAllButton';
+
+type SlidesPerView = 'auto' | 1 | 2 | 3 | 4 | 6;
+
+export enum CarouselPaginationTheme {
+  Dark,
+  Light,
+}
+
+export interface CarouselProps {
+  navigation: boolean;
+  navigationPosition: CarouselNavigationPosition;
+  pagination: boolean;
+  paginationTheme: CarouselPaginationTheme;
+  lgSlidesPerView: SlidesPerView;
+  smSlidesPerView: SlidesPerView;
+  lgSlidesSideOffset: number;
+  smSlidesSideOffset: number;
+  loop: boolean;
+  autoplay: AutoplayOptions | boolean;
+  on?: { [key in SwiperEvent]?: () => void };
+  className?: string;
+  children: React.ReactNode;
+  getSwiper?: (swiper: SwiperInstance) => void;
+}
+
+const DEFAULT_PARAMS = {
+  slideToClickedSlide: true,
+  shouldSwiperUpdate: true,
+  threshold: 10,
+};
+
+const DEFAULT_PAGINATION_PARAMS: { [key in string]: PaginationOptions } = {
+  pagination: {
+    el: '.swiper-pagination',
+    type: 'bullets',
+    clickable: true,
+  },
+};
+
+export default class Carousel extends PureComponent<CarouselProps> {
+  public static ViewAllButton = ViewAllButton;
+  public static defaultProps = {
+    navigation: true,
+    navigationPosition: CarouselNavigationPosition.TopRightOut,
+    pagination: false,
+    paginationTheme: CarouselPaginationTheme.Dark,
+    lgSlidesPerView: 1,
+    smSlidesPerView: 1,
+    lgSlidesSideOffset: 0,
+    smSlidesSideOffset: 0,
+    loop: false,
+    autoplay: false,
+  };
+  private swiper: SwiperInstance = null;
+
+  public render() {
+    const {
+      children,
+      lgSlidesPerView,
+      className,
+      navigation,
+      navigationPosition,
+      paginationTheme,
+      pagination,
+      lgSlidesSideOffset,
+      smSlidesSideOffset,
+    } = this.props;
+
+    const sliderCount = React.Children.count(children);
+    const shouldHideNavigation = typeof lgSlidesPerView === 'number' && sliderCount <= lgSlidesPerView;
+    return (
+      <Container className={className}>
+        <Inner>
+          <SwiperWrapper
+            navigation={navigation}
+            navigationPosition={navigationPosition}
+            pagination={pagination}
+            paginationTheme={paginationTheme}
+            lgSlidesSideOffset={lgSlidesSideOffset}
+            smSlidesSideOffset={smSlidesSideOffset}
+          >
+            <Swiper {...this.getSwiperParams()} getSwiper={this.updateSwiper}>
+              {children}
+            </Swiper>
+          </SwiperWrapper>
+          {navigation && !shouldHideNavigation && (
+            <Navigation position={navigationPosition} goNext={this.goNext} goPrev={this.goPrev} />
+          )}
+        </Inner>
+      </Container>
+    );
+  }
+
+  private updateSwiper = (swiper: SwiperInstance) => {
```

Comment:
updateSwiper라는 이름은 왠지 Swiper를 갱신하는(`swiper.update()`) 행동과 헷갈릴 것 같은데 좋은 이름이 없을까! (getSwiper가 명확하기는 할 것 같은데, getSwiper는 this.props.getSwiper랑 겹쳐서 못 쓴 거겠디..)


그런데 외부에서 getSwiper를 사용해서 swiperInstance를 쓸 일이 있을까?
웬만하면 swiper 자체는 숨겨놓고, 사용해야하는 메서드나 변수만 `<Carousel>` 에서 밖으로 인터페이스를 공개하는 게 좋을 것 같아

Reply:
지금 경우는 external library를 부모 컴포넌트에 전해주는 상황이라 예시가 좀 다르긴 하지만, ref에 대해서도 리액트 문서에서 웬만하면 캡슐화를 하라고 권장하구 있어!
외부에서 swiperjs를 직접 알지 않아도 되게 캡슐화를 잘 해두면, swiperjs에 의존하지 않아도 되기 때문에 나중에 다른 carousel 라이브러리로 교체할 일이 생길 때 고통이 없을 거야

> React components hide their implementation details, including their rendered output. Other components using FancyButton `usually will not need to` obtain a ref to the inner button DOM element. This is good because it prevents components from relying on each other’s DOM structure too much.

e.g. https://reactjs.org/docs/forwarding-refs.html#forwarding-refs-to-dom-components

Reply:
흐으음 그럼 사용하게 될 메서드가 뭐가 될 지 아직은 잘 모르겠다. Infinite 캐러셀도 제거하고 전체보기 버튼으로 가자고 얘기가 되어서! 그럼 필요할 때 추가하기로 할게

Reply:
웅웅 getSwiper prop은 제거하구, 필요할 때 메서드 추가하잣!!

Reply:
아마 swiper.goNext() 나 swiper.currentSlideIndex 같은 게 필요해질듯..!